File: Java/problem1419.java
class Solution {

    private int getIndex(char c) {
        switch (c) {
            case 'c' :
                return 0;
            case 'r':
                return 1;
            case 'o':
                return 2;
            case 'a':
                return 3;
            case 'k':
                return 4;
            default:
                return -1;
        }
    }

    public int minNumberOfFrogs(String croakOfFrogs) {
        char[] chars = croakOfFrogs.toCharArray();
        if (chars.length % 5 != 0) {
            return -1;
        }
        
        int[] counts = new int[5];
        int result = 0;
        for (char c : chars) {
            int idx = getIndex(c);
            ++counts[idx];
            for (int i = 0; i < idx; ++i) {
                if (counts[idx] > counts[i]) {
                    return -1;
                }
            }
            result = Math.max(result, counts[idx]);
            if (idx == 4) {
                for (int i = 0; i < 5; ++i) {
                    --counts[i];
                }
            }
        }
        return result;
    }
}

File: python/problem1419.py
class Solution:
    def minNumberOfFrogs(self, croakOfFrogs: str) -> int:
        if len(croakOfFrogs) % 5 != 0:
            return -1
        char2idx = {'c' : 0, 'r' : 1, 'o' : 2, 'a' : 3, 'k' : 4}
        counts = [0 for _ in range(5)]
        result = 0
        for c in croakOfFrogs:
            idx = char2idx[c]
            counts[idx] += 1
            if idx > 0 and counts[idx] > counts[idx - 1]:
                    return -1
            result = max(counts[idx], result)
            if idx == 4:
                for i in range(5):
                    counts[i] -= 1
        if sum(counts) > 0:
            return -1
        return result
